kube-controller-manager panic on horizonal pod autoscaler asserting "*v1beta2.MetricValueList"

<!-- Please use this template while reporting a bug and provide as much info as possible. Not doing so may result in your bug not being addressed in a timely manner. Thanks!

If the matter is security related, please disclose it privately via https://kubernetes.io/security/
-->


**What happened**:

```
E0717 17:53:09.091799       1 runtime.go:69] Observed a panic: &runtime.TypeAssertionError{interfaceString:"runtime.Object", concreteString:"*v1.Status", assertedString:"*v1beta2.MetricValueList", missingMethod:""} (interface conversion: runtime.Object is *v1.Status, not *v1beta2.MetricValueList)

/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:76
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/asm_amd64.s:573
/usr/local/go/src/runtime/panic.go:502
/usr/local/go/src/runtime/iface.go:252
/usr/local/go/src/runtime/iface.go:262
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/metrics/pkg/client/custom_metrics/versioned_client.go:269
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/metrics/pkg/client/custom_metrics/multi_client.go:136
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/metrics/rest_metrics_client.go:113
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/replica_calculator.go:158
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal.go:347
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal.go:274
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal.go:550
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal.go:318
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal.go:210
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal.go:198
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal.go:164
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134
/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/usr/local/go/src/runtime/asm_amd64.s:2361
panic: interface conversion: runtime.Object is *v1.Status, not *v1beta2.MetricValueList [recovered]
	panic: interface conversion: runtime.Object is *v1.Status, not *v1beta2.MetricValueList

goroutine 1814 [running]:
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x107
panic(0x330f7c0, 0xc4236640c0)
	/usr/local/go/src/runtime/panic.go:502 +0x229
k8s.io/kubernetes/vendor/k8s.io/metrics/pkg/client/custom_metrics.(*namespacedMetrics).GetForObjects(0xc4231d65a0, 0x0, 0x0, 0x3a5b3e8, 0x3, 0x3df7240, 0xc4236c77e0, 0xc422adb460, 0x13, 0x3df72a0, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/metrics/pkg/client/custom_metrics/versioned_client.go:269 +0x4d3
k8s.io/kubernetes/vendor/k8s.io/metrics/pkg/client/custom_metrics.(*multiClientInterface).GetForObjects(0xc422eb9f90, 0x0, 0x0, 0x3a5b3e8, 0x3, 0x3df7240, 0xc4236c77e0, 0xc422adb460, 0x13, 0x3df72a0, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/metrics/pkg/client/custom_metrics/multi_client.go:136 +0x118
k8s.io/kubernetes/pkg/controller/podautoscaler/metrics.(*customMetricsClient).GetRawMetric(0xc4200e1cb0, 0xc422adb460, 0x13, 0xc4204ccf98, 0x7, 0x3df7240, 0xc4236c77e0, 0x3df72a0, 0x5ba7488, 0x0, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/metrics/rest_metrics_client.go:113 +0x102
k8s.io/kubernetes/pkg/controller/podautoscaler.(*ReplicaCalculator).GetMetricReplicas(0xc4215bd980, 0x2, 0x55d4a80, 0xc422adb460, 0x13, 0xc4204ccf98, 0x7, 0x3df7240, 0xc4236c77e0, 0x3df72a0, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/replica_calculator.go:158 +0xb0
k8s.io/kubernetes/pkg/controller/podautoscaler.(*HorizontalController).computeStatusForPodsMetric(0xc420c93680, 0x2, 0xc423e755a0, 0x4, 0x0, 0xc423310ac0, 0x0, 0x0, 0xc423910c40, 0x3df7240, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal.go:347 +0xd8
k8s.io/kubernetes/pkg/controller/podautoscaler.(*HorizontalController).computeReplicasForMetrics(0xc420c93680, 0xc423910c40, 0xc422992500, 0xc423b9a8a0, 0x1, 0x1, 0x11, 0x3b2be6e, 0x3d, 0x0, ...)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal.go:274 +0x933
k8s.io/kubernetes/pkg/controller/podautoscaler.(*HorizontalController).reconcileAutoscaler(0xc420c93680, 0xc4204ae300, 0xc421b5bfe0, 0x17, 0x0, 0x0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal.go:550 +0x16de
k8s.io/kubernetes/pkg/controller/podautoscaler.(*HorizontalController).reconcileKey(0xc420c93680, 0xc421b5bfe0, 0x17, 0x30401e0, 0xc4209576f0, 0xc422f5c140)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal.go:318 +0x277
k8s.io/kubernetes/pkg/controller/podautoscaler.(*HorizontalController).processNextWorkItem(0xc420c93680, 0x16def00)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal.go:210 +0xde
k8s.io/kubernetes/pkg/controller/podautoscaler.(*HorizontalController).worker(0xc420c93680)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal.go:198 +0x2b
k8s.io/kubernetes/pkg/controller/podautoscaler.(*HorizontalController).(k8s.io/kubernetes/pkg/controller/podautoscaler.worker)-fm()
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal.go:164 +0x2a
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc422ed9f60)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x54
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc422ed9f60, 0x3b9aca00, 0x0, 0x1, 0xc4200b0780)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xbd
k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.Until(0xc422ed9f60, 0x3b9aca00, 0xc4200b0780)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
created by k8s.io/kubernetes/pkg/controller/podautoscaler.(*HorizontalController).Run
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/controller/podautoscaler/horizontal.go:164 +0x1ca
```

**What you expected to happen**: kube-controller-manager should not panic.

**How to reproduce it (as minimally and precisely as possible)**: Kubernetes 1.12.6, `k8s.gcr.io/cluster-autoscaler:v1.3.4`, and https://github.com/zalando-incubator/kube-metrics-adapter

**Anything else we need to know?**:

It's running `k8s.gcr.io/cluster-autoscaler:v1.3.4` and https://github.com/zalando-incubator/kube-metrics-adapter, and `kube-metrics-adapter` logs says:

```
I0717 16:19:54.651255       1 handler.go:143] kube-metrics-adapter: GET "/apis/custom.metrics.k8s.io/v1beta1/namespaces/default/pods/*/total-response-rate" satisfied by gorestful with webservice /apis/custom.metrics.k8s.io/v1beta1
E0717 16:19:54.651414       1 writers.go:149] apiserver was unable to write a JSON response: expected pointer, but got nil
E0717 16:19:54.651427       1 status.go:64] apiserver received an error that is not an metav1.Status: &errors.errorString{s:"expected pointer, but got nil"}
I0717 16:19:54.652574       1 wrap.go:42] GET /apis/custom.metrics.k8s.io/v1beta1/namespaces/default/pods/%2A/total-response-rate?
...
```

Don't see any notable errors in `cluster-autoscaler`.

Is that because it's running the old `cluster-autoscaler`?

Basically, https://github.com/kubernetes/metrics/blob/13eb3b486029dcf3189111977d0865a933d92c28/pkg/client/custom_metrics/versioned_client.go#L234-L237 is expecting `*v1beta2.MetricValueList`, but somehow it's getting `*v1.Status` which I assume `"k8s.io/apimachinery/pkg/apis/meta/v1.Status"` https://github.com/kubernetes/apimachinery/blob/bfcf53abc9f82bad3e534fcb1c36599d3c989ebf/pkg/apis/meta/v1/types.go#L582.

**Environment**:
- Kubernetes version (use `kubectl version`): v1.12.6
- Cloud provider or hardware configuration: AWS EC2
- OS (e.g: `cat /etc/os-release`): Amazon Linux 2
- Kernel (e.g. `uname -a`): 4.14.109-99.92.amzn2.x86_64

--- a/staging/src/k8s.io/metrics/pkg/client/custom_metrics/versioned_client.go
+++ b/staging/src/k8s.io/metrics/pkg/client/custom_metrics/versioned_client.go
@@ -18,11 +18,12 @@ package custom_metrics
 
 import (
 	"fmt"
+	"reflect"
 
 	"k8s.io/apimachinery/pkg/api/meta"
 	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/apimachinery/pkg/runtime/schema"
-	serializer "k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
 	"k8s.io/client-go/rest"
 	"k8s.io/client-go/util/flowcontrol"
 
@@ -122,7 +123,11 @@ func (m *rootScopedMetrics) getForNamespace(namespace string, metricName string,
 		return nil, err
 	}
 
-	res := metricObj.(*v1beta2.MetricValueList)
+	var res *v1beta2.MetricValueList
+	var ok bool
+	if res, ok = metricObj.(*v1beta2.MetricValueList); !ok {
+		return nil, fmt.Errorf("the custom metrics API server didn't return MetricValueList, the type is %v", reflect.TypeOf(metricObj))
+	}
 	if len(res.Items) != 1 {
 		return nil, fmt.Errorf("the custom metrics API server returned %v results when we asked for exactly one", len(res.Items))
 	}
@@ -160,7 +165,11 @@ func (m *rootScopedMetrics) GetForObject(groupKind schema.GroupKind, name string
 		return nil, err
 	}
 
-	res := metricObj.(*v1beta2.MetricValueList)
+	var res *v1beta2.MetricValueList
+	var ok bool
+	if res, ok = metricObj.(*v1beta2.MetricValueList); !ok {
+		return nil, fmt.Errorf("the custom metrics API server didn't return MetricValueList, the type is %v", reflect.TypeOf(metricObj))
+	}
 	if len(res.Items) != 1 {
 		return nil, fmt.Errorf("the custom metrics API server returned %v results when we asked for exactly one", len(res.Items))
 	}
@@ -199,7 +208,11 @@ func (m *rootScopedMetrics) GetForObjects(groupKind schema.GroupKind, selector l
 		return nil, err
 	}
 
-	res := metricObj.(*v1beta2.MetricValueList)
+	var res *v1beta2.MetricValueList
+	var ok bool
+	if res, ok = metricObj.(*v1beta2.MetricValueList); !ok {
+		return nil, fmt.Errorf("the custom metrics API server didn't return MetricValueList, the type is %v", reflect.TypeOf(metricObj))
+	}
 	return res, nil
 }
 
@@ -234,7 +247,11 @@ func (m *namespacedMetrics) GetForObject(groupKind schema.GroupKind, name string
 		return nil, err
 	}
 
-	res := metricObj.(*v1beta2.MetricValueList)
+	var res *v1beta2.MetricValueList
+	var ok bool
+	if res, ok = metricObj.(*v1beta2.MetricValueList); !ok {
+		return nil, fmt.Errorf("the custom metrics API server didn't return MetricValueList, the type is %v", reflect.TypeOf(metricObj))
+	}
 	if len(res.Items) != 1 {
 		return nil, fmt.Errorf("the custom metrics API server returned %v results when we asked for exactly one", len(res.Items))
 	}
@@ -269,6 +286,10 @@ func (m *namespacedMetrics) GetForObjects(groupKind schema.GroupKind, selector l
 		return nil, err
 	}
 
-	res := metricObj.(*v1beta2.MetricValueList)
+	var res *v1beta2.MetricValueList
+	var ok bool
+	if res, ok = metricObj.(*v1beta2.MetricValueList); !ok {
+		return nil, fmt.Errorf("the custom metrics API server didn't return MetricValueList, the type is %v", reflect.TypeOf(metricObj))
+	}
 	return res, nil
 }

